qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,409 @@
+        rpm2cpio = subprocess.Popen(['rpm2cpio', path], stdout=subprocess.PIPE)
+        # `-D` is GNUism
+        cpio = subprocess.Popen([
+                'cpio',
+                '-idm',
+                '-D',
+                target,
+                '.%s/%s/*' % (PATH_PREFIX, name)
+            ], stdin=rpm2cpio.stdout, stdout=subprocess.DEVNULL)
+        return rpm2cpio.wait() == 0 and cpio.wait() == 0
+
+def parse_config(path):
+    with open(path, 'r') as f:
+        return dict(line.rstrip('\n').split('=', 1) for line in f)
+
+def install(args, app):
+    # TODO: Lock, mentioned in the note above
+    transaction_set = rpm.TransactionSet()
+
+    rpm_list = []
+    for template in args.templates:
+        if template.endswith('.rpm'):
+            if not os.path.exists(template):
+                parser.error('RPM file \'%s\' not found.' % template)
+            rpm_list.append(template)
+
+    os.makedirs(args.cachedir, exist_ok=True)
+
+    dl_list = get_dl_list(args, app)
+    dl_list_copy = dl_list.copy()
+    # Verify that the templates are not yet installed
+    for name, ver in dl_list.items():
+        if name in app.domains:
+            print(('Template \'%s\' already installed, skipping...'
+                ' (You may want to use the {reinstall,upgrade,downgrade}'
+                ' operations.)') % name, file=sys.stderr)
+            del dl_list_copy[name]
+        else:
+            version_str = build_version_str(ver)
+            target_file = \
+                '%s%s-%s.rpm' % (PACKAGE_NAME_PREFIX, name, version_str)
+            rpm_list.append(os.path.join(args.cachedir, target_file))
+    dl_list = dl_list_copy
+
+    download(args, app, path_override=args.cachedir, dl_list=dl_list)
+
+    for rpmfile in rpm_list:
+        if not verify_rpm(rpmfile, args.nogpgcheck, transaction_set):
+            parser.error('Package \'%s\' verification failed.' % template)
+
+    for rpmfile in rpm_list:
+        with tempfile.TemporaryDirectory(dir=TEMP_DIR) as target:
+            package_hdr = get_package_hdr(rpmfile)
+            package_name = package_hdr[rpm.RPMTAG_NAME]
+            if not package_name.startswith(PACKAGE_NAME_PREFIX):
+                parser.error(
+                    'Illegal package name for package \'%s\'.' % rpmfile)
+            # Remove prefix to get the real template name
+            name = package_name[len(PACKAGE_NAME_PREFIX):]
+
+            # Another check for already-downloaded RPMs
+            if name in app.domains:
+                print(('Template \'%s\' already installed, skipping...'
+                    ' (You may want to use the {reinstall,upgrade,downgrade}'
+                    ' operations.)') % name, file=sys.stderr)
+                continue
+
+            print('Installing template \'%s\'...' % name, file=sys.stderr)
+            extract_rpm(name, rpmfile, target)
+            cmdline = [
+                'qvm-template-postprocess',
+                '--keep-source',
```

Comment:
Why `--keep-source` (besides debugging)?

Reply:
Oops, that was indeed something I did for debugging. As TemporaryDirectory removes the directory, I did not notice this while testing it afterwards.